### ISSUE TITLE
[HOLD for payment 2024-02-07] [$500] IOU - Replacing Receipt with an Invalid File Format Does Not Display an Error Message

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#28343]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login to any account
2. Send a Money Request to another account, using the Scan Receipt option
3. Click on the Receipt image to open the preview popup
4. Click on the “Three dots” icon >> select the “Replace” option
5. Replace the current image with another image in an unsupported format, such as "webp"
6. Observe that: there is no error, but when reloading the page, the old image still appears (which indicates that new image is not uploaded)

## Expected Result:
In this case, the user has uploaded an invalid image format, so there should be an error message. The file should remain optimistically in Onyx until the user decides to clear the error.

## Actual Result:
No error displays.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.74-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/166e4fc0-399f-46db-a228-ddb051b597fb

https://github.com/Expensify/App/assets/78819774/f6ced923-c723-4a5b-ad67-41c893991a9a

![8123857_img-0539_720](https://github.com/Expensify/App/assets/78819774/f0774f3b-5831-481c-aa33-f21f8a7c0769)

https://github.com/Expensify/App/assets/78819774/a7664b82-b7b2-435d-9135-354db35680cc

**Expensify/Expensify Issue URL:**

**Issue reported by:** @tranvantoan-qn

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695720594688229

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bb8307379994bc8e</li>
        <li>Upwork Job ID: 1707109060385632256</li>
        <li>Last Price Increase: 2023-10-18</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Reviewer | 27267319</li>
<li>tienifr | Contributor | 27267321</li>
<li>tranvantoan-qn | Reporter | 27267322</li>
</ul></li>
    </ul>
</details>

**Comment by ikevin127** [comment]
# Proposal
## Please re-state the problem that we are trying to solve in this issue
The app allows the user to upload an invalid file format but then shows the old image.

## What is the root cause of that problem?
The problem stems from the `validateReceipt` function located in the `ReceiptSelector/index.js` component: 
```
function validateReceipt(file) {
    const {fileExtension} = FileUtils.splitExtensionFromFileName(lodashGet(file, 'name', ''));
    if (_.contains(CONST.API_ATTACHMENT_VALIDATIONS.UNALLOWED_EXTENSIONS, fileExtension.toLowerCase())) {
        setUploadReceiptError(true, 'attachmentPicker.wrongFileType', 'attachmentPicker.notAllowedExtension');
        return false;
    }
    ...
```

In this case `UNALLOWED_EXTENSION` does not exist within `CONST.API_ATTACHMENT_VALIDATIONS`, therefore no file format is excluded from upload as receipt.

## What changes do you think we should make in order to solve the problem?
Would add `UNALLOWED_EXTENSION` constant as an array of comma separated strings with the not-allowed formats (TBD).

## What alternative solutions did you explore? (Optional)
N/A

Example with `UNALLOWED_EXTENSION: ['pdf'];`
## Videos

https://github.com/Expensify/App/assets/56457735/34739481-0e0c-4040-b63e-ce9752d1148a

**Comment by trjExpensify** [comment]
@JmillsExpensify @youssef-lr, how does this one jive with [this issue](https://github.com/Expensify/App/issues/11189) for allowing all file extensions? Seems like we're silently failing in this case. 

CC: @AndrewGable @Gonals as well from the receipt upload implementation

**Comment by trjExpensify** [comment]
>Currently, if we fail to upload a receipt, although BE already returns an error, the App won't revert changes of optimisticData and show the previous receipt until we reload the page.

Hm, that sounds correct to me. We should show an error message immediately, but we shouldn't revert the receipt until they take action on the error message to dismiss it. For example.. 

1. I'm offline on a plane, and decide to do my expenses on the ride home 
2. I upload a receipt for $100 to get paid back from my company, and I throw the receipt away
3. We optimistically create the request with the receipt
4. I go back online and the receipt upload fails for some reason
5. I don't have the receipt anymore, so I download it, and try again

If we just reverted the receipt that failed to upload, a user could very well be out of pocket, so I think we need to be careful with that. 

CC: @mountiny @iwiznia

**Comment by iwiznia** [comment]
Yeah I think I agree with that. We need to provide a way for you to get the thing you uploaded

**Comment by trjExpensify** [comment]
Yeah, so I think that should apply with a failed receipt upload when replacing a receipt here? We display the error message, and when you dismiss it, it reverts back to the original.

**Comment by tranvantoan-qn** [comment]
So, does that mean we will now check the file format locally and show an error if it is not a supported file type, instead of sending it to the backend immediately?

**Comment by trjExpensify** [comment]
I think we should fix it properly, yeah. Otherwise people will lose receipts that failed to be replaced. 

@Gonals might have some insight on the approach for this from the Scan Receipt implementation, as we should be doing the same to give them the option to download it on a failed receipt upload.

**Comment by jasperhuangg** [comment]
So it sounds like the given proposal that was approved by C+ doesn't entirely fit our requirements. We don't want to lose the optimistic file in Onyx that was uploaded, in case the user wants to redownload the file and try again.

This is still open for proposals!

**Comment by jasperhuangg** [comment]
@hoangzinh done

**Comment by ikevin127** [comment]
[Proposal](https://github.com/Expensify/App/issues/28343#issuecomment-1738165307) updated after latest **Expected result** changes.

**Comment by trjExpensify** [comment]
>In this case, the user has uploaded an invalid image format, so there should be an error message. The file should remain optimistically in Onyx until the user decides to clear the error.

@jasperhuangg we should make this clear that they should maintain an option to download it as well, right? Otherwise it's still lost when they clear the error. 

As for the in-line error message, I would think it follows this structure: 

>The new receipt didn’t upload, [save the file](download.com) to not lose it, or dismiss this error to revert back to the original receipt.    
>[insert API error message returned here]

Clicking "save the file" in the error message downloads a copy of the failed file upload.

**Comment by iwiznia** [comment]
> @jasperhuangg we should make this clear that they should maintain an option to download it as well, right? Otherwise it's still lost when they clear the error.

hmmm yeah there's a big risk if we clear the file with clearing the error, if the customer does not save it first, but also we probably can't keep it forever there, right? What if we made the button to dismiss the error ALSO trigger a download? That way we ensure the user saved it (or had the possibility of saving it). We can make it clear in the error that they should save the file when closing the x, if it's possible that the file was not saved (not sure if it is).

**Comment by trjExpensify** [comment]
>What if we made the button to dismiss the error ALSO trigger a download?

That seems like a good way to make sure it's not overlooked if we can. 

>We can make it clear in the error that they should save the file when closing the x, if it's possible that the file was not saved (not sure if it is).

I'm not sure it is either, wouldn't it just go to downloads? 

New error message proposal then being something like..

>The new receipt didn’t upload. Dismissing this error will download a copy of the file and revert back to the original receipt.
>[insert API error message returned here]

**Comment by trjExpensify** [comment]
Hm, on iOS I'd expect it to save to `Files`, not sure on Android though.

**Comment by trjExpensify** [comment]
>There are some concerns in my mind about the "Dismissing this error will download a copy of the file". I think this function only has advantages if user uploads a receipt via Taking a photo. Other methods like uploading a receipt => they already had it stored in their device => If we auto download, we might save another duplicated receipt in their device.

Saving a dupe is better than losing it if it's a trade-off, but I get the point. They might have a copy so don't need to download an additional one, so by making it auto-download when dismissing the error message we eliminate the choice presuming they don't have a copy. In which case perhaps the hyperlinked [save the file] button is better so they have that choice. 

>Moreover, in order to do it, on some platforms users need to grant permission to save a file.

Sure, but that's the case either way we allow or make them take the action to save the file? 

>I have an idea, instead of downloading, is it better if we have a button/icon/text to re-upload/retry?

I don't think so. Retrying over and over doesn't help you get a copy of the receipt back.

>Last but not least, I feel like either downloading or re-uploading are kind of feature request rather than fixing a bug. Just wanna confirm again if we would like to do either of them in this issue. Thanks

I guess you could argue that it's a feature request, but we're looking to solve the bug the right way here. Happy to make this issue a `new feature` if that's the consensus, but let's fix it properly.

**Comment by jasperhuangg** [comment]
I personally like the idea of giving the user the choice of whether to redownload the file or not. To me, dismissing the error to download the file doesn't seem very clear to me, and it could also be confusing to the user especially if they're given a duplicated file.

Why would we need to redownload the file if we dismiss the error though? Wouldn't the user always have the file if they uploaded it?

**Comment by trjExpensify** [comment]
That could work. Though do you think it's a bit hidden? Like are you going to be scared to hit the `X` not knowing that you'll then have an option to download it? Or are you thinking it's..

In-line:
>The new receipt didn’t upload, [download the file](https://github.com/Expensify/App/issues/download.com) to not lose it, or dismiss this error to revert back to the original receipt.
>[insert API error message returned here]

If they tap the X to dismiss the error then they get an additional modal to confirm:

>**Are you sure?**
>Download the receipt first if you don't have another copy of it. 
>[Delete]
>[Cancel]

**Comment by jasperhuangg** [comment]
I like that second option more, I feel like in-line links would be difficult to tap on mobile. But I think we should probably bring this to a wider audience since this is really feeling like a new feature at this point.

**Comment by trjExpensify** [comment]
Ah, it wouldn't be either one or two, but both. To illustrate where we're at with it: 

![image](https://github.com/Expensify/App/assets/16232057/b29f1a4f-df6e-469c-9eac-1edecf551a18)

Happy to take this to open-source, though probably tomorrow at this point!

**Comment by jasperhuangg** [comment]
Ohh I see what you mean @trjExpensify sorry for the confusion. With the mockups it looks better to me. Let's taken it to #expensify-open-source to get some more eyes on it. Did you get these mockups from design (are these mockups design approved)?

**Comment by trjExpensify** [comment]
Thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1696502156555529

> Did you get these mockups from design (are these mockups design approved)?

Nah, I created them myself to illustrate the proposal!

**Comment by trjExpensify** [comment]
Okay, so wrapping up after that convo. We're aligned, and looking to proceed with this: 

![image](https://github.com/Expensify/App/assets/16232057/e0410d27-1f8b-45ef-a0fa-2434a7bd1dba)

The copy has been edited ever so slightly!

**Comment by trjExpensify** [comment]
@jasperhuangg can you please do a secondary review of the proposal please? Thanks!

**Comment by trjExpensify** [comment]
@jasperhuangg still waiting on your secondary view of the proposal.

**Comment by trjExpensify** [comment]
Bumped @jasperhuangg

**Comment by trjExpensify** [comment]
> I found out that, after uploading the pdf file, BE server still returns success data, is it correct?

PDFs are allowed for receipt uploads. Interestingly, there's a bug with them displaying though that we're talking about [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1697672309044349?thread_ts=1696936253.355669&cid=C01GTK53T8Q).  

>I also want to confirm that which extensions that are not allowed.

Same with this one in the thread. We have a list of allowed extensions in the BE, but the FE check needs adding that Alberto is looking into. Jump into the thread :)

**Comment by trjExpensify** [comment]
Ah also, I just realised that @Gonals is still working on the OG implementation of this in the request creation flow in [this PR](https://github.com/Expensify/App/pull/29790#issuecomment-1770502754). Let's make sure we're all aligned here for complimentary work. :)

**Comment by trjExpensify** [comment]
Yeah cool, makes sense to me.

**Comment by trjExpensify** [comment]
The linked PR we're holding on is still in review.

**Comment by trjExpensify** [comment]
Still in review, conflicts for @Gonals to address!

**Comment by trjExpensify** [comment]
comments have been addressed on the linked PR we're holding on. Awaiting a re-review from @0xmiroslav! :)

**Comment by jasperhuangg** [comment]
This is still on HOLD cc @0xmiroslav

**Comment by trjExpensify** [comment]
No change, linked PR is still in review.

**Comment by trjExpensify** [comment]
Linked PR still in review. C+ approved, awaiting the secondary review(s) now from @marcochavezf & @nkuoch.

**Comment by trjExpensify** [comment]
Let's get moving!

**Comment by trjExpensify** [comment]
Have you gone through the replace receipt flow? Here's the expected for reference: https://github.com/Expensify/App/issues/28343#issuecomment-1749823256

**Comment by trjExpensify** [comment]
@Gonals your PR was (partially?) reverted a couple of weeks ago and I can't see where it was addressed to reimplement. Can you jump back on this one please?

**Comment by trjExpensify** [comment]
@Gonals has put a PR in review to unblock you here: https://github.com/Expensify/App/pull/32562

**Comment by trjExpensify** [comment]
@0xmiroslav can you review https://github.com/Expensify/App/pull/32562 please?

**Comment by trjExpensify** [comment]
Still held. Commented in [the PR](https://github.com/Expensify/App/pull/32562#issuecomment-1854767329).

**Comment by jasperhuangg** [comment]
Still held on the PR, looks like good progress is being made in the review.

**Comment by jasperhuangg** [comment]
Same

**Comment by trjExpensify** [comment]
@hoangzinh we should be unblocked here now! :)

**Comment by trjExpensify** [comment]
@jasperhuangg over to you for the secondary review!

**Comment by jasperhuangg** [comment]
@tienifr let's move forward with your [proposal](https://github.com/Expensify/App/issues/28343#issuecomment-1738414257)!

**Comment by trjExpensify** [comment]
👋 checklist time here!

**Comment by trjExpensify** [comment]
Nice, thanks! Applause have a regression test for the failed upload side, so I've created an issue for the replace receipt side.

**Comment by trjExpensify** [comment]
Summarising payouts as follows: 

- $500 to @hoangzinh for C+ review
- $500 to @tienifr for the fix
- $50 to @tranvantoan-qn for the bug report (prior to the Oct 24th cut off).

Settled up, thanks!
